File: 3rd-year/ca341-comparative-programming-languages/comparative-programming-assignment/imperative-solution/NameBinaryTree.c
#include <stdlib.h>
#include "string.h"
#include "./NameBinaryTree.h"

char* nameSearch(struct NameNode *root, char *name)
{
    if(root == 0)
    {
        char *answer = malloc(10 * 100 * sizeof(char));
        strcpy(answer, "not ");
        strcat(answer, "found");
        return answer;
    }
    if(strcmp(name, root->name) == 0)
    {
        char *answer = malloc(10 * 100 * sizeof(char));
        strcpy(answer, "found : ");
        strcat(answer, root->name);
        strcat(answer, " : ");
        strcat(answer, root->number);
        strcat(answer, " : ");
        strcat(answer, root->address);
        return answer;
    }
    if(strcmp(name, root->name) < 0)
        nameSearch(root->left, name);
    else if(strcmp(name, root->name) > 0)
        nameSearch(root->right, name);
}

void inorderTraversalName(struct NameNode *root)
{
    if(root == 0)
        return;
    inorderTraversalName(root->left);
    if(strcmp(root->name, "ROOT") != 0)
        printf("%s : %s : %s \n", root->name, root->number, root->address);
    inorderTraversalName(root->right);
}

struct NameNode* newNameNode(char *name, char *number, char *address)
{
    //struct NameNode *temp = (struct NameNode *)malloc(sizeof(struct NameNode));
    struct NameNode *temp = malloc(sizeof(struct NameNode));
    temp->name = name;
    temp->number = number;
    temp->address = address;
    temp->left = temp->right = 0;
    return temp;
}

struct NameNode* addNameNode(struct NameNode *root, char *name, char *number, char *address)
{
    if(root == 0)
        return newNameNode(name, number, address);
    if(strcmp(name, root->name) < 0)
        root->left = addNameNode(root->left, name, number, address);
    else if(strcmp(name, root->name) > 0)
        root->right = addNameNode(root->right, name, number, address);
    else
        printf("error: duplicate in tree");
    return root;
}

struct NameNode* minValueNameNode(struct NameNode  *root)
{
    if(root->left == 0)
        return root;
    return minValueNameNode(root->left);
}

struct NameNode* removeNameNode(struct NameNode *root, char *name)
{
    if(root == NULL)
    {
        return root;
    }
    else if(strcmp(name, root->name) < 0)
    {
        root->left = removeNameNode(root->left, name);
    }
    else if(strcmp(name, root->name) > 0)
    {
        root->right = removeNameNode(root->right, name);
    }
    else
    {
        if(root->left == 0)
        {
            struct NameNode *temp = root->right;
            free(root);
            return temp;
        }
        if(root->right == 0)
        {
            struct NameNode *temp = root->left;
            free(root);
            return temp;
        }
        struct NameNode *temp = minValueNameNode(root->right);
        root->name = name;
        root->right = removeNameNode(root->right, temp->name);
    }
    return root;
}

File: 3rd-year/ca341-comparative-programming-languages/comparative-programming-assignment/oop-solution/NameBinaryTree.py
class NameNode:
    def __init__(self, name, number, address):
        self.name = name
        self.number = number
        self.address = address
        self.left = None
        self.right = None

class NameBinaryTree:
    def __init__(self):
        self.root = None

    def search(self, name):
        # iterative implementation
        current_node = self.root
        while current_node != None and current_node.name != name:
            if name < current_node.name:
                current_node = current_node.left
            elif name > current_node.name:
                current_node = current_node.right
            else:
                raise Exception("error: duplicate in tree")
        if current_node != None and current_node.name == name:
            return [True, current_node.name, current_node.number, current_node.address]
        else:
            return False

    def inorderTraversal(self):
        self.r_inorderTraversal(self.root)

    def r_inorderTraversal(self, node):
        if node == None:
            return
        self.r_inorderTraversal(node.left)
        print(node.name + " : " + node.number + " : " + node.address)
        self.r_inorderTraversal(node.right)

    def add(self, name, number, address):
        self.root = self.r_add(self.root, name, number, address)

    def r_add(self, root, name, number, address):
        if root == None:
            self.root = NameNode(name, number, address)
            return self.root
        if name < root.name:
            root.left = self.r_add(root.left, name, number, address)
        elif name > root.name:
            root.right = self.r_add(root.right, name, number, address)
        else:
            raise Exception("error: duplicate in tree")
        return root

    def min_value_node(self, node):
        current = node
        # loop down to find the leftmost leaf
        while(current.left is not None):
            current = current.left
        return current

    def remove(self, name):
        self.root = self.r_remove(self.root, name)

    def r_remove(self, root, name):
        if root == None:
            return root
        if name < root.name:
            root.left = self.r_remove(root.left, name)
        elif name > root.name:
            root.right = self.r_remove(root.right, name)
        else:
            # Node with only one child or no child
            if root.left is None:
                temp = root.right
                root = None
                return temp

            if root.right is None:
                temp = root.left
                root = None
                return temp

            # If the node has two children, replace it with the smallest
            # node in its right subtree
            temp = self.min_value_node(root.right)
            root.name = temp.name
            root.right = self.r_remove(root.right, temp.name)
        return root
